File: OJ practices/LeetCode题目集/117. Populating Next Right Pointers in Each Node II(medium)-层序遍历法.java
/*
 * @Author: SourDumplings
 * @Date: 2020-09-21 09:22:34
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/populating-next-right-pointers-in-each-node-ii/
 */

class Solution
{
    public Node connect(Node root)
    {
        Queue<Node> q = new LinkedList<Node>();
        if (root == null)
        {
            return null;
        }
        q.add(root);
        while (!q.isEmpty())
        {
            List<Node> list = new ArrayList<Node>();
            while (!q.isEmpty())
            {
                list.add(q.remove());
            }
            int s = list.size();
            for (int i = 0; i < s - 1; i++)
            {
                Node thisNode = list.get(i);
                thisNode.next = list.get(i + 1);
                if (thisNode.left != null)
                {
                    q.add(thisNode.left);
                }
                if (thisNode.right != null)
                {
                    q.add(thisNode.right);
                }
            }
            Node lastNode = list.get(s - 1);
            list.get(s - 1).next = null;
            if (lastNode.left != null)
            {
                q.add(lastNode.left);
            }
            if (lastNode.right != null)
            {
                q.add(lastNode.right);
            }
        }
        return root;
    }
}

File: OJ practices/LeetCode题目集/117. Populating Next Right Pointers in Each Node II(medium)-层序遍历法.cpp
/**
 * @file 117. Populating Next Right Pointers in Each Node II(medium)-层序遍历法.cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief
 * @version 1.0.0
 * @date 2024-12-09
 *
 * @copyright Copyright (c) 2024 SourDumplings
 *
 */

/*
// Definition for a Node.
class Node {
public:
	int val;
	Node* left;
	Node* right;
	Node* next;

	Node() : val(0), left(NULL), right(NULL), next(NULL) {}

	Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

	Node(int _val, Node* _left, Node* _right, Node* _next)
		: val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution
{
public:
	Node* connect(Node* root)
	{
		if (root == nullptr) { return nullptr; }

		vector<Node*> thisLevelNodes;
		thisLevelNodes.push_back(root);
		while (!thisLevelNodes.empty())
		{
			size_t n = thisLevelNodes.size();
			vector<Node*> nextLevelNodes;
			for (size_t i = 0; i < n; ++i)
			{
				Node* node = thisLevelNodes[i];
				if (i < n - 1) { node->next = thisLevelNodes[i + 1]; }
				if (node->left) { nextLevelNodes.push_back(node->left); }
				if (node->right) { nextLevelNodes.push_back(node->right); }
			}
			thisLevelNodes = std::move(nextLevelNodes);
		}
		return root;
	}
};